src/software/simulated_tests/ball_moves_in_direction.py

```diff
@@ -0,0 +1,137 @@
+        :param moving_in_pos_x: True if checking for +x direction, false if checking for -x direction
+        """
+        self.last_ball_position = initial_ball_position
+        self.moving_in_pos_x = moving_in_pos_x
+        self.max_displacement_so_far = None
+
+    def get_validation_status(self, world) -> ValidationStatus:
+        """Checks if ball is moving forward or backward
+
+        :param world: The world msg to validate
+        :returns: FAILING if ball doesn't move in the direction
+                  PASSING if ball moves in the direction
+        """
+        current_ball_position = tbots.createPoint(
+            world.ball.current_state.global_position
+        ).x()
+
+        # if max displacement is not set or current ball is moving in the right direction
+        # set it and return PASSING
+        if self.max_displacement_so_far is None or (
+            (self.moving_in_pos_x and current_ball_position > self.max_displacement_so_far)
+            or (
+                not self.moving_in_pos_x
+                and current_ball_position < self.max_displacement_so_far
+            )
+        ):
+            self.max_displacement_so_far = current_ball_position
+            return ValidationStatus.PASSING
+
+        # if max displacement is set and current ball is in the wrong direction too far
+        # beyond a threshold, return FAILING
+        if (
+            self.moving_in_pos_x
+            and current_ball_position < self.max_displacement_so_far - 0.1
+        ) or (
+            not self.moving_in_pos_x
+            and current_ball_position > self.max_displacement_so_far + 0.1
+        ):
+            return ValidationStatus.FAILING
+
+    def get_validation_geometry(self, world) -> ValidationGeometry:
+        """
+        (override) Shows the last ball position line
+        """
+        return create_validation_geometry(
+            [
+                tbots.Rectangle(
+                    tbots.Point(
+                        self.last_ball_position.x(),
+                        tbots.Field(world.field).fieldBoundary().yMin(),
+                    ),
+                    tbots.Point(
+                        self.last_ball_position.x() + 0.01,
+                        tbots.Field(world.field).fieldBoundary().yMax(),
+                    ),
+                )
+            ]
+        )
+
+    def __repr__(self):
+        return (
+            "Check that the ball moves " + "forward" if self.moving_in_pos_x else "backward"
+        )
+
+
+class BallMovesForward(BallMovesXAxis):
+
+    """Checks if ball is moving forward, i.e. in the +x direction"""
+
+    def __init__(self, initial_ball_position):
+        super().__init__(initial_ball_position, True)
+
+
+(
+    BallEventuallyMovesForward,
+    BallEventuallyStopsMovingForward,
+    BallAlwaysMovesForward,
+    BallNeverMovesForward,
+) = create_validation_types(BallMovesForward)
+
+
+class BallMovesInDirectionInRegions(BallMovesXAxis):
```

Comment:
why wouldn't you have two validation functions: `BallMovesInDirection` and `RobotsInRegion` in the eventually/aggregate validation set

Reply:
bump